[Doran] Vending app step2

hwj0623 opened the pull request:
step2 내용 구현 했습니다.
- CollectionView 사용했습니다.
   - 음료수 이미지/버튼/Label은 하나의 CollectionCell로 묶어처리하였습니다.
- 학습거리의 ImageView 커스터마이징 구현
- 재고추가 버튼에 대한 예외처리 Alert으로 NotificationCenter를 사용해보았습니다.

감사합니다.

hwj0623 committed:
자판기App 스토리보드 기본틀 작성
- 이미지 리소스추가

hwj0623 committed:
컬렉션뷰 및 음료수 이미지 추가

hwj0623 committed:
자판기 잔액 충전 버튼기능 추가

hwj0623 committed:
메뉴테이블 크기와 특정 음료수의 재고크기 구하는 메서드 추가

hwj0623 committed:
재고 추가 기능 구현
- 시스템오류 Alert을 NotificationCenter에 Observer추가
* Delegate, DataSource 메서드 구현 분리
* ImageView에 대한 둥근 테두리 구현

hwj0623 committed:
step2 구현내용에 대해 정리 및 alert 메시지수정

hwj0623 committed:
참고 레퍼런스 사이트 링크 추가

godrm commented on VendingMachineApp/VendingMachineApp/VendingViewController.swift:
```diff
@@ -0,0 +1,130 @@
+            chargeInput = 1000
+        }else if balanceSize == "+5000"{
+            chargeInput = 5000
+        }
```
값을 비교할 때 title 문자열과 비교하는 것은 좋지 않습니다.
표시하는 문자는 다국어를 지원할 수 있거나 양식이 바뀔 수 있습니다. 
버튼 객체 인스턴스를 비교하는 게 좋습니다. 다른 방법을 고민해보세요. 
메뉴를 매칭하는 방식을 같이 고민해보세요. 
그리도 문자열이나 숫자도 하드코딩 하기 보다는 코드로 표현하세요.

hwj0623 replied:
버튼 태그 비교방식으로 개선하겠습니다.
 - restorationIdentifier로 설정하려고 했는데, 데이터타입이 역시 String이라 이슈가 생길 수 있겠네요

godrm commented on VendingMachineApp/VendingMachineApp/VendingViewController.swift:
```diff
@@ -0,0 +1,130 @@
+
+class GridCell: UICollectionViewCell{
+    @IBOutlet var imgView: UIImageView!
+    @IBOutlet weak var addButton: UIButton!
```
cell 객체는 별도 파일로 분리하세요.

godrm commented on VendingMachineApp/VendingMachineApp/VendingViewController.swift:
```diff
@@ -0,0 +1,130 @@
+    
+    @IBAction func addDrinkStockButton(_ sender: UIButton) {
+        do {
+            let drink = try vendingMachine.selectProduct(productId: itemIndex)
```
cell을 사용할 때는 주의할 점이 있습니다.
늘 반복해서 재사용된다는 점입니다. 
그리고 vendingMachine 직접 접근하는 것은 좋지 않습니다. 
버튼 액션을 만들꺼면 아래처럼 Notification을 보내서 ViewController가 처리하는 게 좋습니다.

hwj0623 replied:
TO BE 로직을 세우다 보니 뭔가 어색한게 있어서 한가지 질문드리고자 합니다. 

**AS IS** : cell 내에서 버튼이벤트 발생 -> GridCell 클래스 내에서 `자판기 재고 갱신` 및 `Label text값` 갱신  -> 실패시 NotificationCenter로 post -> `ViewController에서 전달받은 에러데이터를 기반으로 에러 핸들러 함수 실행(alert창 띄우기)`

**TO BE** : cell 내에서 버튼이벤트 발생 -> cell의 인덱스를 NotificationCenter로 전달 (post)-> ViewController에서 NotificationCenter로부터 전달받은 index로 `자판기 재고 갱신` -> 실패시 NotificationCenter로 전달 -> ViewController에서 재차 NotificationCenter로부터 전달받은 에러데이터 기반으로 에러 핸들러 함수 실행
- 이러면 Label text의 값 갱신을 위해 특정 sender(UIButton)를 식별하여 값을 재차 전달해야 하는 로직에 구멍이 생겨버립니다. 버튼으로부터 전달받은 index를 기반으로 버튼을 다시 찾아서 Label을 갱신해야 하는가요?

hwj0623 replied:
요약하자면 ViewController가 NotificationCenter로부터 값을 제대로 받았는지, 관찰대상인 sender에게 알릴 방법이 없는지 궁금하네요.

godrm commented on VendingMachineApp/VendingMachineApp/VendingViewController.swift:
```diff
@@ -0,0 +1,130 @@
+    private func setImageViewBorderRound(){
+        imgView.layer.cornerRadius = 10
+        imgView.layer.borderWidth = 1
+        imgView.layer.borderColor = UIColor.cyan.cgColor
```
이미지 뷰를 구체적으로 바꾸는 동작을 한다면 ImageView를 상속하는 서브뷰를 만들고 여기 코드를 그 내부로 옮기는 게 좋습니다. 
뷰 객체도 모델처럼 구체적인 세부 동작은 뷰 객체 내부로 옮기는 게 좋습니다.

godrm commented on VendingMachineApp/VendingMachineApp/VendingViewController.swift:
```diff
@@ -0,0 +1,130 @@
+        }
+        stockInfoLabel.text = "\(drinkItemList.drinkStockList.count)개"
+        itemIndex = index+1
+        self.vendingMachine = vendingMachine
```
뷰에 값을 설정할 때는 모든것을 다 넘기기 보다 뷰를 설정하기 위한 최소한의 데이터만 넘기는 게 좋습니다. 어떤 데이터가 필요한지 고민해보세요.

hwj0623 replied:
재고 size, item index를 나타내는 Int값 두개면 충분하네요. 개선하겠습니다

hwj0623 committed:
GridCell 분리 및 버튼 식별기능 수정
- MoneyQuantity 단위를 위한 enum 추가

hwj0623 committed:
fix : 재고추가 기능 개선
- GridCell에서 vendingmachine 다루지 않고 Notification거치도록 변경

hwj0623 committed:
GridCell 내 vendingMachine 프로퍼티 제거
- 인덱스 보정 로직 변경 (gridcell 내에서 -> collectionView 메서드 호출시점으로)

hwj0623 committed:
UIimageView subclass로 분리
- 이미지 크기조절 실패

hwj0623 committed:
이미지 안뜨는 오류 수정
- storyboard의 UIImageView는 커스텀 서브클래스로 지정
- 커스텀 서브클래스인 DrinkUIImageView의 layoutSubviews 정의
   - 해당 메서드는 오버라이드 생성자 이후에 호출됨
- GridCell에서 imgView의 데이터 타입은 UIImageView로 설정, image 프로퍼티만 UIImage생성해서 할당

hwj0623 committed:
imgView의 데이터타입을 하위 서브클래스로 변경

hwj0623 commented:
피드백 주신 내용 전부 수정해서 반영하였습니다. 
감사합니다.

godrm commented on VendingMachineApp/Enum/MoneyQuantity.swift:
```diff
@@ -0,0 +1,23 @@
+        case .fiveThousand:
+            return "\(self.rawValue)"
+        }
+    }
```
돈 단위를 위해서 타입으로 만드는 것도 기존 보다 좋지만,
돈 자체를 추상화하는 게 더 좋을 수 있습니다. 참고하세요.

godrm commented on VendingMachineApp/VendingMachineApp/VendingViewController.swift:
```diff
@@ -14,13 +14,41 @@ class VendingViewController: UIViewController{
+        guard let itemIndex = object as? Int else{
+            NotificationCenter.default.post(
+                name: .addDrinkButtonError,
+                object: VendingMachineError.notFoundDrinkIdError
```
ViewController 객체 내부에서 post 하는게 의미가 있을까요? 
이 부분은 다시 한 번 생각해보세요.

godrm commented:
콜랙션뷰를 사용하는 것은 이번 스탭의 요구사항을 벗어나기 때문에 깊이 리뷰하지 않았습니다.
다만 셀을 재사용하는 게 있어서 불필요한 요소는 꼭 개선해야 해서 피드백했습니다. 
다음 단계를 하면서 노티를 보내는 구조도 다시 개선해보세요.
